pystac/summaries.py

```diff
@@ -0,0 +1,262 @@
+import os
+import json
+import numbers
+import urllib.request
+from enum import Enum
+
+from pystac.utils import get_required
+
+from typing import (
+    Any,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Type,
+    Union,
+    cast,
+    TypeVar,
+    Iterable,
+    Protocol,
+    TYPE_CHECKING
+)
+
+if TYPE_CHECKING:
+    from pystac.item import Item as Item_Type
+    from pystac.collection import Collection as Collection_Type
+
+from abc import abstractmethod
+
+
+class Comparable(Protocol):
+    """Protocol for annotating comparable types."""
+
+    @abstractmethod
+    def __lt__(self: "T", other: "T") -> bool:
+        pass
+
+
+T = TypeVar("T", bound=Comparable)
+
+
+class RangeSummary(Generic[T]):
+    def __init__(self, minimum: T, maximum: T):
+        self.minimum = minimum
+        self.maximum = maximum
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {"minimum": self.minimum, "maximum": self.maximum}
+
+    def update_with_value(self, v: T) -> None:
+        self.minimum = min(self.minimum, v)
+        self.maximum = max(self.maximum, v)
+
+    @classmethod
+    def from_dict(cls, d: Dict[str, Any], typ: Type[T] = Any) -> "RangeSummary[T]":
+        minimum: Optional[T] = get_required(d.get("minimum"), "RangeSummary", "minimum")
+        maximum: Optional[T] = get_required(d.get("maximum"), "RangeSummary", "maximum")
+        return cls(minimum=minimum, maximum=maximum)
+
+
+FIELDS_JSON_URL = "https://cdn.jsdelivr.net/npm/@radiantearth/stac-fields/fields.json"
+
+FIELDS_JSON_LOCAL_PATH = os.path.join(os.path.dirname(__file__), "resources",
+                                      "fields.json")
+
+
+class SummaryStrategy(Enum):
+    ARRAY = "v"
+    RANGE = "r"
+    SCHEMA = "s"
+    DONT_SUMMARIZE = False
+    DEFAULT = True
+
+
+class Summarizer():
+    '''The Summarizer computes summaries from values, following the definition of fields
+    to summarize provided in a json file.
+
+    For more information about the structure of the fields json file, see:
+
+    https://github.com/stac-utils/stac-fields
+
+    Args:
+        fields(str): the path to the json file with field descriptions.
+        If no file is passed, a default one will be used.
+    '''
+
+    __default_field_definitions: dict = {}
+
+    def __init__(self, fields: Optional[str] = None):
+        if fields is None:
+            self._set_default_field_definitions()
+        else:
+            try:
+                with open(fields) as f:
+                    jsonfields = json.load(f)
+                self._set_field_definitions(jsonfields)
+            except:
+                self._set_default_field_definitions()
```

Comment:
I feel like this should throw the error and not result in a silent fallback to the default - if someone were to pass in fields and have expectations that they would be used, this could be a very tricky thing to debug from results downstream